In [134]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

In [135]:
%matplotlib inline

In [136]:
df = pd.read_excel('Data_Train.xlsx')

In [137]:
df.head()

Name    Location  Year  Kilometers_Driven  \
0            Maruti Wagon R LXI CNG      Mumbai  2010              72000   
1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
2                      Honda Jazz V     Chennai  2011              46000   
3                 Maruti Ertiga VDI     Chennai  2012              87000   
4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0       CNG       Manual      First  26.6 km/kg   998 CC  58.16 bhp    5.0   
1    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
2    Petrol       Manual      First   18.2 kmpl  1199 CC   88.7 bhp    5.0   
3    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
4    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   

   Price  
0   1.75  
1  12.50  
2   4.50  
3   6.00  
4  17.74

In [138]:
df.tail()

Name   Location  Year  Kilometers_Driven Fuel_Type  \
6014          Maruti Swift VDI      Delhi  2014              27365    Diesel   
6015  Hyundai Xcent 1.1 CRDi S     Jaipur  2015             100000    Diesel   
6016     Mahindra Xylo D4 BSIV     Jaipur  2012              55000    Diesel   
6017        Maruti Wagon R VXI    Kolkata  2013              46000    Petrol   
6018     Chevrolet Beat Diesel  Hyderabad  2011              47000    Diesel   

     Transmission Owner_Type     Mileage   Engine     Power  Seats  Price  
6014       Manual      First   28.4 kmpl  1248 CC    74 bhp    5.0   4.75  
6015       Manual      First   24.4 kmpl  1120 CC    71 bhp    5.0   4.00  
6016       Manual     Second   14.0 kmpl  2498 CC   112 bhp    8.0   2.90  
6017       Manual      First   18.9 kmpl   998 CC  67.1 bhp    5.0   2.65  
6018       Manual      First  25.44 kmpl   936 CC  57.6 bhp    5.0   2.50

In [139]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6019 entries, 0 to 6018
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               6019 non-null   object 
 1   Location           6019 non-null   object 
 2   Year               6019 non-null   int64  
 3   Kilometers_Driven  6019 non-null   int64  
 4   Fuel_Type          6019 non-null   object 
 5   Transmission       6019 non-null   object 
 6   Owner_Type         6019 non-null   object 
 7   Mileage            6017 non-null   object 
 8   Engine             5983 non-null   object 
 9   Power              5983 non-null   object 
 10  Seats              5977 non-null   float64
 11  Price              6019 non-null   float64
dtypes: float64(2), int64(2), object(8)
memory usage: 564.4+ KB


We can see some null values in Mileage, Engine, Power and Seats.

In [140]:
df.describe()

Year  Kilometers_Driven        Seats        Price
count  6019.000000       6.019000e+03  5977.000000  6019.000000
mean   2013.358199       5.873838e+04     5.278735     9.479468
std       3.269742       9.126884e+04     0.808840    11.187917
min    1998.000000       1.710000e+02     0.000000     0.440000
25%    2011.000000       3.400000e+04     5.000000     3.500000
50%    2014.000000       5.300000e+04     5.000000     5.640000
75%    2016.000000       7.300000e+04     5.000000     9.950000
max    2019.000000       6.500000e+06    10.000000   160.000000

In [141]:
df['Fuel_Type'].unique()

array(['CNG', 'Diesel', 'Petrol', 'LPG', 'Electric'], dtype=object)

In [142]:
df[df['Fuel_Type'] == 'Electric']

Name Location  Year  Kilometers_Driven Fuel_Type  \
4446       Mahindra E Verito D4  Chennai  2016              50000  Electric   
4904  Toyota Prius 2009-2016 Z4   Mumbai  2011              44000  Electric   

     Transmission Owner_Type Mileage   Engine   Power  Seats  Price  
4446    Automatic      First     NaN    72 CC  41 bhp    5.0  13.00  
4904    Automatic      First     NaN  1798 CC  73 bhp    5.0  12.75

## Cleaning the Data

Removing null values.

In [143]:
new_df = df.dropna()

In [144]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5975 entries, 0 to 6018
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               5975 non-null   object 
 1   Location           5975 non-null   object 
 2   Year               5975 non-null   int64  
 3   Kilometers_Driven  5975 non-null   int64  
 4   Fuel_Type          5975 non-null   object 
 5   Transmission       5975 non-null   object 
 6   Owner_Type         5975 non-null   object 
 7   Mileage            5975 non-null   object 
 8   Engine             5975 non-null   object 
 9   Power              5975 non-null   object 
 10  Seats              5975 non-null   float64
 11  Price              5975 non-null   float64
dtypes: float64(2), int64(2), object(8)
memory usage: 606.8+ KB


After removing all the nulls we see that there are 5975 rows left in the Dataframe.

In [145]:
new_df.describe()

Year  Kilometers_Driven        Seats        Price
count  5975.000000       5.975000e+03  5975.000000  5975.000000
mean   2013.386778       5.867431e+04     5.278828     9.501647
std       3.247238       9.155851e+04     0.808959    11.205736
min    1998.000000       1.710000e+02     0.000000     0.440000
25%    2012.000000       3.390800e+04     5.000000     3.500000
50%    2014.000000       5.300000e+04     5.000000     5.650000
75%    2016.000000       7.300000e+04     5.000000     9.950000
max    2019.000000       6.500000e+06    10.000000   160.000000

In [146]:
new_df['Fuel_Type'].unique()

array(['CNG', 'Diesel', 'Petrol', 'LPG'], dtype=object)

In [147]:
new_df[new_df['Seats'] == 0]

Name   Location  Year  Kilometers_Driven  \
3999  Audi A4 3.2 FSI Tiptronic Quattro  Hyderabad  2012             125000   

     Fuel_Type Transmission Owner_Type    Mileage   Engine     Power  Seats  \
3999    Petrol    Automatic      First  10.5 kmpl  3197 CC  null bhp    0.0   

      Price  
3999   18.0

In [148]:
new_df['Seats'].unique()

array([ 5.,  7.,  8.,  4.,  6.,  2., 10.,  9.,  0.])

In [149]:
new_df['Owner_Type'].unique()

array(['First', 'Second', 'Fourth & Above', 'Third'], dtype=object)

In [150]:
new_df[new_df['Power'] == 'null bhp']

Name   Location  Year  \
76                Ford Fiesta 1.4 SXi TDCi     Jaipur  2008   
79                  Hyundai Santro Xing XL  Hyderabad  2005   
89                  Hyundai Santro Xing XO  Hyderabad  2007   
120   Hyundai Santro Xing XL eRLX Euro III     Mumbai  2005   
143    Hyundai Santro Xing XO eRLX Euro II      Kochi  2008   
...                                    ...        ...   ...   
5861                Hyundai Santro Xing XO    Chennai  2007   
5873   Hyundai Santro Xing XO eRLX Euro II       Pune  2006   
5925           Skoda Laura Classic 1.8 TSI       Pune  2010   
5943               Mahindra Jeep MM 540 DP    Chennai  2002   
5985                       Toyota Etios GD      Delhi  2013   

      Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage  \
76               111111    Diesel       Manual      First   17.8 kmpl   
79                87591    Petrol       Manual      First    0.0 kmpl   
89                73745    Petrol       Manual      First   17.0 kmpl   
120              102000    Petrol       Manual     Second   17.0 kmpl   
143               80759    Petrol       Manual      Third   17.0 kmpl   
...                 ...       ...          ...        ...         ...   
5861              79000    Petrol       Manual      First   17.0 kmpl   
5873              47200    Petrol       Manual     Second   17.0 kmpl   
5925              85000    Petrol       Manual      First   17.5 kmpl   
5943              75000    Diesel       Manual      First    0.0 kmpl   
5985              70000    Diesel       Manual      First  23.59 kmpl   

       Engine     Power  Seats  Price  
76    1399 CC  null bhp    5.0   2.00  
79    1086 CC  null bhp    5.0   1.30  
89    1086 CC  null bhp    5.0   2.10  
120   1086 CC  null bhp    5.0   0.85  
143   1086 CC  null bhp    5.0   1.67  
...       ...       ...    ...    ...  
5861  1086 CC  null bhp    5.0   1.85  
5873  1086 CC  null bhp    5.0   1.20  
5925  1798 CC  null bhp    5.0   2.85  
5943  2112 CC  null bhp    6.0   1.70  
5985  1364 CC  null bhp    5.0   3.88  

[103 rows x 12 columns]

In [151]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5975 entries, 0 to 6018
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               5975 non-null   object 
 1   Location           5975 non-null   object 
 2   Year               5975 non-null   int64  
 3   Kilometers_Driven  5975 non-null   int64  
 4   Fuel_Type          5975 non-null   object 
 5   Transmission       5975 non-null   object 
 6   Owner_Type         5975 non-null   object 
 7   Mileage            5975 non-null   object 
 8   Engine             5975 non-null   object 
 9   Power              5975 non-null   object 
 10  Seats              5975 non-null   float64
 11  Price              5975 non-null   float64
dtypes: float64(2), int64(2), object(8)
memory usage: 606.8+ KB


In [152]:
new_df.describe()

Year  Kilometers_Driven        Seats        Price
count  5975.000000       5.975000e+03  5975.000000  5975.000000
mean   2013.386778       5.867431e+04     5.278828     9.501647
std       3.247238       9.155851e+04     0.808959    11.205736
min    1998.000000       1.710000e+02     0.000000     0.440000
25%    2012.000000       3.390800e+04     5.000000     3.500000
50%    2014.000000       5.300000e+04     5.000000     5.650000
75%    2016.000000       7.300000e+04     5.000000     9.950000
max    2019.000000       6.500000e+06    10.000000   160.000000

In [153]:
new_df = new_df[new_df['Power']!='null bhp']

In [154]:
new_df.describe()

Year  Kilometers_Driven        Seats        Price
count  5872.000000       5.872000e+03  5872.000000  5872.000000
mean   2013.477691       5.831700e+04     5.283719     9.603919
std       3.164568       9.216941e+04     0.805081    11.249453
min    1998.000000       1.710000e+02     2.000000     0.440000
25%    2012.000000       3.342250e+04     5.000000     3.517500
50%    2014.000000       5.260900e+04     5.000000     5.750000
75%    2016.000000       7.240275e+04     5.000000    10.000000
max    2019.000000       6.500000e+06    10.000000   160.000000

In [155]:
new_df["Power"]= new_df["Power"].str.split(" ", n=1, expand = True) 


In [156]:
new_df["Engine"]= new_df["Engine"].str.split(" ", n=1, expand = True) 
new_df["Mileage"]= new_df["Mileage"].str.split(" ", n=1, expand = True) 

In [157]:
new_df.tail()

Name   Location  Year  Kilometers_Driven Fuel_Type  \
6014          Maruti Swift VDI      Delhi  2014              27365    Diesel   
6015  Hyundai Xcent 1.1 CRDi S     Jaipur  2015             100000    Diesel   
6016     Mahindra Xylo D4 BSIV     Jaipur  2012              55000    Diesel   
6017        Maruti Wagon R VXI    Kolkata  2013              46000    Petrol   
6018     Chevrolet Beat Diesel  Hyderabad  2011              47000    Diesel   

     Transmission Owner_Type Mileage Engine Power  Seats  Price  
6014       Manual      First    28.4   1248    74    5.0   4.75  
6015       Manual      First    24.4   1120    71    5.0   4.00  
6016       Manual     Second    14.0   2498   112    8.0   2.90  
6017       Manual      First    18.9    998  67.1    5.0   2.65  
6018       Manual      First   25.44    936  57.6    5.0   2.50

In [158]:
new_df["Power"] = pd.to_numeric(new_df["Power"], downcast="float")
new_df["Engine"] = pd.to_numeric(new_df["Engine"], downcast="float")
new_df["Mileage"] = pd.to_numeric(new_df["Mileage"], downcast="float")

In [159]:
new_df.tail()

Name   Location  Year  Kilometers_Driven Fuel_Type  \
6014          Maruti Swift VDI      Delhi  2014              27365    Diesel   
6015  Hyundai Xcent 1.1 CRDi S     Jaipur  2015             100000    Diesel   
6016     Mahindra Xylo D4 BSIV     Jaipur  2012              55000    Diesel   
6017        Maruti Wagon R VXI    Kolkata  2013              46000    Petrol   
6018     Chevrolet Beat Diesel  Hyderabad  2011              47000    Diesel   

     Transmission Owner_Type    Mileage  Engine       Power  Seats  Price  
6014       Manual      First  28.400000  1248.0   74.000000    5.0   4.75  
6015       Manual      First  24.400000  1120.0   71.000000    5.0   4.00  
6016       Manual     Second  14.000000  2498.0  112.000000    8.0   2.90  
6017       Manual      First  18.900000   998.0   67.099998    5.0   2.65  
6018       Manual      First  25.440001   936.0   57.599998    5.0   2.50

In [160]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5872 entries, 0 to 6018
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               5872 non-null   object 
 1   Location           5872 non-null   object 
 2   Year               5872 non-null   int64  
 3   Kilometers_Driven  5872 non-null   int64  
 4   Fuel_Type          5872 non-null   object 
 5   Transmission       5872 non-null   object 
 6   Owner_Type         5872 non-null   object 
 7   Mileage            5872 non-null   float32
 8   Engine             5872 non-null   float32
 9   Power              5872 non-null   float32
 10  Seats              5872 non-null   float64
 11  Price              5872 non-null   float64
dtypes: float32(3), float64(2), int64(2), object(5)
memory usage: 527.6+ KB


In [170]:
new_df.describe()

Year  Kilometers_Driven      Mileage       Engine        Power  \
count  5872.000000       5.872000e+03  5872.000000  5872.000000  5872.000000   
mean   2013.477691       5.831700e+04    18.277824  1625.745605   113.277344   
std       3.164568       9.216941e+04     4.365658   601.639343    53.881828   
min    1998.000000       1.710000e+02     0.000000   624.000000    34.200001   
25%    2012.000000       3.342250e+04    15.260000  1198.000000    75.000000   
50%    2014.000000       5.260900e+04    18.200001  1495.500000    97.699997   
75%    2016.000000       7.240275e+04    21.100000  1991.000000   138.100006   
max    2019.000000       6.500000e+06    33.540001  5998.000000   560.000000   

             Seats        Price  
count  5872.000000  5872.000000  
mean      5.283719     9.603919  
std       0.805081    11.249453  
min       2.000000     0.440000  
25%       5.000000     3.517500  
50%       5.000000     5.750000  
75%       5.000000    10.000000  
max      10.000000   160.000000

In [161]:
new_df.to_excel('new_train.xlsx')

In [162]:
df_test = pd.read_excel('Data_Test.xlsx')

In [163]:
df_test.head()

Name    Location  Year  \
0                    Maruti Alto K10 LXI CNG       Delhi  2014   
1              Maruti Alto 800 2016-2019 LXI  Coimbatore  2013   
2  Toyota Innova Crysta Touring Sport 2.4 MT      Mumbai  2017   
3                       Toyota Etios Liva GD   Hyderabad  2012   
4                          Hyundai i20 Magna      Mumbai  2014   

   Kilometers_Driven Fuel_Type Transmission Owner_Type      Mileage   Engine  \
0              40929       CNG       Manual      First  32.26 km/kg   998 CC   
1              54493    Petrol       Manual     Second    24.7 kmpl   796 CC   
2              34000    Diesel       Manual      First   13.68 kmpl  2393 CC   
3             139000    Diesel       Manual      First   23.59 kmpl  1364 CC   
4              29000    Petrol       Manual      First    18.5 kmpl  1197 CC   

       Power  Seats  
0   58.2 bhp    4.0  
1   47.3 bhp    5.0  
2  147.8 bhp    7.0  
3   null bhp    5.0  
4  82.85 bhp    5.0

In [164]:
df_test = df_test.dropna()

In [165]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1223 entries, 0 to 1233
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               1223 non-null   object 
 1   Location           1223 non-null   object 
 2   Year               1223 non-null   int64  
 3   Kilometers_Driven  1223 non-null   int64  
 4   Fuel_Type          1223 non-null   object 
 5   Transmission       1223 non-null   object 
 6   Owner_Type         1223 non-null   object 
 7   Mileage            1223 non-null   object 
 8   Engine             1223 non-null   object 
 9   Power              1223 non-null   object 
 10  Seats              1223 non-null   float64
dtypes: float64(1), int64(2), object(8)
memory usage: 114.7+ KB


In [166]:
df_test = df_test[df_test['Power']!='null bhp']

In [167]:
df_test["Power"]= df_test["Power"].str.split(" ", n=1, expand = True) 
df_test["Engine"]= df_test["Engine"].str.split(" ", n=1, expand = True) 
df_test["Mileage"]= df_test["Mileage"].str.split(" ", n=1, expand = True) 


In [168]:
df_test["Power"] = pd.to_numeric(df_test["Power"], downcast="float")
df_test["Engine"] = pd.to_numeric(df_test["Engine"], downcast="float")
df_test["Mileage"] = pd.to_numeric(df_test["Mileage"], downcast="float")

In [169]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1201 entries, 0 to 1233
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               1201 non-null   object 
 1   Location           1201 non-null   object 
 2   Year               1201 non-null   int64  
 3   Kilometers_Driven  1201 non-null   int64  
 4   Fuel_Type          1201 non-null   object 
 5   Transmission       1201 non-null   object 
 6   Owner_Type         1201 non-null   object 
 7   Mileage            1201 non-null   float32
 8   Engine             1201 non-null   float32
 9   Power              1201 non-null   float32
 10  Seats              1201 non-null   float64
dtypes: float32(3), float64(1), int64(2), object(5)
memory usage: 98.5+ KB


In [171]:
df_test.to_excel('new_test.xlsx')

In [175]:
new_df['Location'].unique()

array(['Mumbai', 'Pune', 'Chennai', 'Coimbatore', 'Hyderabad', 'Jaipur',
       'Kochi', 'Kolkata', 'Delhi', 'Bangalore', 'Ahmedabad'],
      dtype=object)